server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -109,598 +109,510 @@ protected void masterOperation(
+            request.isSingleRepositoryRequest() == false,
             request.snapshots(),
             request.ignoreUnavailable(),
-            request.verbose(),
-            (CancellableTask) task,
+            SnapshotPredicates.fromRequest(request),
             request.sort(),
-            request.after(),
-            request.offset(),
-            request.size(),
             request.order(),
             request.fromSortValue(),
-            SnapshotPredicates.fromRequest(request),
-            request.includeIndexNames(),
-            listener
-        );
+            request.offset(),
+            request.after(),
+            request.size(),
+            SnapshotsInProgress.get(state),
+            request.verbose(),
+            request.includeIndexNames()
+        ).getMultipleReposSnapshotInfo(listener);
     }
 
     /**
-     * Filters the list of repositories that a request will fetch snapshots from in the special case of sorting by repository
-     * name and having a non-null value for {@link GetSnapshotsRequest#fromSortValue()} on the request to exclude repositories outside
-     * the sort value range if possible.
+     * A single invocation of the get-snapshots API.
      */
-    private static List<RepositoryMetadata> maybeFilterRepositories(
-        List<RepositoryMetadata> repositories,
-        GetSnapshotsRequest.SortBy sortBy,
-        SortOrder order,
-        @Nullable String fromSortValue
-    ) {
-        if (sortBy != GetSnapshotsRequest.SortBy.REPOSITORY || fromSortValue == null) {
-            return repositories;
-        }
-        final Predicate<RepositoryMetadata> predicate = order == SortOrder.ASC
-            ? repositoryMetadata -> fromSortValue.compareTo(repositoryMetadata.name()) <= 0
-            : repositoryMetadata -> fromSortValue.compareTo(repositoryMetadata.name()) >= 0;
-        return repositories.stream().filter(predicate).toList();
-    }
-
-    private void getMultipleReposSnapshotInfo(
-        boolean isMultiRepoRequest,
-        SnapshotsInProgress snapshotsInProgress,
-        TransportGetRepositoriesAction.RepositoriesResult repositoriesResult,
-        String[] snapshots,
-        boolean ignoreUnavailable,
-        boolean verbose,
-        CancellableTask cancellableTask,
-        GetSnapshotsRequest.SortBy sortBy,
-        @Nullable GetSnapshotsRequest.After after,
-        int offset,
-        int size,
-        SortOrder order,
-        String fromSortValue,
-        SnapshotPredicates predicates,
-        boolean indices,
-        ActionListener<GetSnapshotsResponse> listener
-    ) {
-        // Process the missing repositories
-        final Map<String, ElasticsearchException> failures = ConcurrentCollections.newConcurrentMap();
-        for (String missingRepo : repositoriesResult.missing()) {
-            failures.put(missingRepo, new RepositoryMissingException(missingRepo));
+    private class GetSnapshotsOperation {
```

Comment:
We could, but then basically nothing would be left in `TransportGetSnapshotsAction` and we'd still have all that size and complexity in one place. And indeed one could argue it's kinda more complex as a top-level class because that means the reader has to worry about who else in the package might be calling it, whereas as a private inner class we can see all the callers right in front of us.

Reply:
After talking offline, I think this is more of a lack of strong API boundaries problem. Ideally there's an API that says a method does X, and callers can use it as needed. But the GetSnapshotsOperation namespace isn't really a logical piece, rather a namespace we're dumping everything into until we refactor more meaningful boundaries. For the time being.

Multiple callers means deduplicating code, which sounds like a good problem.